### PR TITLE
feat: add file watcher service

### LINKED ISSUE #1683: epic: Cortex track delete models on file system
### Cortex version

1.0.2

### Describe the issue and expected behaviour

- Problem: Jan users are used to deleting the model folder, rather than deleting on Jan UI
- Users cannot just remove the /models folder as the backend has to update cortex.db
- We should track the file system and update cortex.db so that no action is required for them

Approaches (@louis-jan & @namchuai to add on)
- Build cortex.db as a quick lookup table, but like caching, we still monitor folder changes for better performance.
- Scan it once on launch in the background
- To propose filescan interval
- Alternative: `model has been deleted! -> delete in cortex.db`

## QA:
- Delete model folder (or specifically model.yml file)
- expected result: cortex models list does not list the model

### PR DESCRIPTION
**Pull request by namchuai**
## Describe Your Changes

- Add file watcher service

## Fixes Issues

- fixes #1683 

## Self Checklist

- [ ] Added relevant comments, esp in complex areas
- [ ] Updated docs (for bug fixes / features)
- [ ] Created issues for follow-up changes or refactoring needed

### CODE REVIEW

File: engine/services/file_watcher_service.h [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+#endif
+    running = false;
+    if (watchThread.joinable()) {
+      watchThread.join();
+    }
+    CTL_INF("FileWatcherService stopped!");
+  }
+
+ private:
+  std::string watchPath;
```
Comment (vansangpfiev) [inline engine/services/file_watcher_service.h]: Maybe should rename to `watch_path` for coding convention consistent.

File: engine/services/file_watcher_service.h [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    fd = inotify_init();
+    if (fd < 0) {
+      throw std::runtime_error("Failed to initialize inotify");
+    }
+
+    // Add initial watch on the main directory
+    AddWatch(watchPath);
+
+    char buffer[4096];
+    while (running) {
```
Comment (vansangpfiev) [inline engine/services/file_watcher_service.h]: Should we put a sleep in the while loop?

File: engine/services/file_watcher_service.h [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    watchThread = std::thread(&FileWatcherService::watcherThread, this);
+  }
+
+  void stop() {
+#ifdef _WIN32
+    CloseHandle(dir_handle);
+#endif
+
+#ifdef Linux
+    CleanupWatches();
```
Comment (vansangpfiev) [inline engine/services/file_watcher_service.h:102]: I think we might have a race condition for `watchDescriptors`

File: engine/services/file_watcher_service.h (hunk 2 above)
Comment (namchuai) [reply engine/services/file_watcher_service.h]: `read()` is a blocking call. I don't think we should add sleep in there.

**Review by vansangpfiev — APPROVED** [review]

**Pull request merged**